### DISCUSSION TITLE
perplexities about `triplets()` method in DimeNet class

### DISCUSSION THREAD

**Discussion by fedeotto** [pyg-team/pytorch_geometric#4953] [Q&A]
I know somebody already asked for clarifications about this function works, but I'm trying to understand what `triplets` does at a more basic level. I understand that for each directional message `m_ji` from source index `j` (`row`) to target index `i` (`col`) we need to consider messages `m_kj` i.e. all neighbors of the _fixed_ neighbor `j` of target node `i`. Let's say that we have a simple graph with `4` nodes and connectivity:

```
edge_index  :  tensor([[1, 0, 2, 3, 1, 1],
                       [0, 1, 1, 1, 2, 3]])
```

I understand the assignment `row, col = edge_index  # j->i` (source to target).

Then (I suppose), an adjacency matrix is created by doing:

`adj_t = SparseTensor(row=col, col=row, value=value,
                     sparse_sizes=(num_nodes, num_nodes))`

why do we transpose row and column here? Also, when we do:

```
adj_t_row = adj_t[row]

num_triplets = adj_t_row.set_value(None).sum(dim=1).to(torch.long) 
```

What is the exact meaning of taking `adj_t[row]` in this context? And why `num_triplets` is calculated on top of this?

Many thanks in advance and apologize if this was a bit trivial to ask.

**Comment by rusty1s** [comment]
`adj_t` refers to the transposed adjacency matrix of `(row, col)` such that we can index by destination in the first dimension.
Then, `adj_t` will replicate the adjacencies for every entry in `row` (it will return a `[num_edges, num_nodes]` sparse matrix). If we sum this matrix up, then we know the number of triplets (for each edge, we sum up its adjacent nodes).

I agree that the function is very difficult to understand. I suggest to check the output one by one via a simple example.
